### ISSUE TITLE
Set compiler flags depending on used C(++) compiler during Python build

### ISSUE THREAD

**Issue by dweindl** [ICB-DCM/AMICI#812]
Required for MSVC++ support.

See problem here: https://github.com/ICB-DCM/AMICI/pull/807#issuecomment-535576706 

See potential solution here: https://stackoverflow.com/questions/28641408/how-to-tell-which-compiler-will-be-invoked-for-a-python-c-extension-in-setuptool

**Comment by thomassligon** [comment]
The potential solution looks to me like a dirty hack. Do we really need that? How does Python & PIP discover the compiler today? After all, it called MSVC (cl.exe) correctly on my machine, and it was just the extra options that weren't correct. It looks to me like Python already knows which compiler is installed, or finds out somewhere else. Or maybe it's AMICI code.

For example, running a command cl -? returns
Microsoft (R) C/C++ Optimizing Compiler Version 19.22.27905 for x64
(and a lot more lines).
There must be something similar for GCC.

Another example would be this C++ program
`#include <iostream>

int main()
{
#ifdef \_MSC\_VER
	std::cout << "MSVC\n";
#endif;
#ifdef \_\_GNUC\_\_
	std::cout << "GNUV\n";
#endif;
#ifdef \_\_STDC\_\_
	std::cout << "STDC\n";
#endif;
	return 0;
}`

**Comment by dweindl** [comment]
All correct. But you can't use the preprocessor to pass compiler flags, this needs to happen before, i.e. in Python when invoking the compiler.

setuptools discovers the compiler, but then one still needs to change all arguments depending on that result. Would not call that 'dirty', but rather 'annoying'. Way easier if we would use cmake (#721) right away...

**Comment by thomassligon** [comment]
Writing different arguments once depending on an existing flag sounds easy to me, just a few lines, and certainly easier than the stackoverflow post that suggests creating C++ files on the fly and compiling them to see if the compiler creates an error message, then parsing the message.

On the other hand, I recently used existing sundials CMAKE on Visual Studio, and that works very nicely in the current version, but I haven't written any CMAKE for over 30 years and I didn't know much about it then. A quick look seems to be telling me that the compiler options are embedded in the CMAKE files, so I don't see how that would make things easier.

**Comment by dweindl** [comment]
> Writing different arguments once depending on an existing flag sounds easy to me, just a few lines, [...]

Not saying it's rocket science. Simply it will take me longer than half an hour without providing any obvious benefit, as things are working well enough with mingw. Contributions are welcome though.

**Comment by thomassligon** [comment]
There are benefits for me personally.
When searching for an IDE for Python on the web, many reviews don't even mention Visual Studio. However, this one has some statistics
[https://codecondo.com/best-ides-and-code-editors-for-python-developer-2019/](https://codecondo.com/best-ides-and-code-editors-for-python-developer-2019/)
showing that Visual Studio is number 2 at 12%, whereas Visual Studio Code (an open source editor for Windows, Mac and Linux) leads with 20%. PyCharm comes in 9th at 5%.
Both Visual Studio Code (including many third-party add-ons) and Visual Studio Community are free of charge.
Over the past 10 years, I have used Visual Studio to develop in F#, C#, C, C++, and Python. By continuing to use it, I have one IDE, including editor, compiler, and debugger, for all of the languages. The Python compiler and packages are treated as an add-on to Visual Studio, and are not developed by Microsoft. Even though I am not actively developing very much in any of these languages, it is convenient to have the same tool for all of them. In addition, even though I have never used it, and maybe never will, Visual Studio supports debugging in all languages and on multiple machines, including Linux, in a single session. It should be possible to debug a Windows client application and a Linux server backend in one session.
According to Wikipedia, Microsoft's first C compiler was released in 1983, and Linux was invented in 1991. According to Microsoft, compatibility of Visual Studio C++ with the ISO standard and Linux was addressed (unfortunately very late) in VS 2015, with very big improvements in 2017 and more in 2019.
[https://docs.microsoft.com/en-us/cpp/overview/cpp-conformance-improvements?view=vs-2019&viewFallbackFrom=2019](https://docs.microsoft.com/en-us/cpp/overview/cpp-conformance-improvements?view=vs-2019&viewFallbackFrom=2019)
The only problems we have with MSVC in AMICI are trivial things like one line of GNU-specific code or one compiler option.
